psm_utils/io/idxml.py

```diff
@@ -13,33 +13,116 @@
 
 from __future__ import annotations
 
+from abc import abstractmethod
+from typing import Iterable, Tuple, List
 import re
+from pathlib import Path
 
-from pyteomics.openms import idxml
+import pyopenms as oms
 
-from psm_utils.io._base_classes import ReaderBase
+from psm_utils.io._base_classes import ReaderBase, WriterBase
 from psm_utils.psm import PSM
 from psm_utils.psm_list import PSMList
+from psm_utils.exceptions import PSMUtilsException
 
 # Patterns to match open and closed round/square brackets
 MOD_PATTERN = re.compile(r"\(((?:[^)(]+|\((?:[^)(]+|\([^)(]*\))*\))*)\)")
 MOD_PATTERN_NTERM = re.compile(r"^\.\[((?:[^][]+|\[(?:[^][]+|\[[^][]*\])*\])*)\]")
 MOD_PATTERN_CTERM = re.compile(r"\.\[((?:[^][]+|\[(?:[^][]+|\[[^][]*\])*\])*)\]$")
 
+# Extracted from the OpenMS PSMFeatureExtractor, which adds and manipulates features that will be given to percolator
+# https://github.com/OpenMS/OpenMS/blob/342f6524e76a2bab3dcb428ba2f4aa2d6bfe8483/src/topp/PSMFeatureExtractor.cpp
+RESOCRING_FEATURE_LIST = [
+    "isotope_error"
+    "MS:1002049",  # MSGFPlus unchanged RawScore
+    "MS:1002050",  # MSGFPlus unchanged DeNovoScore
+    "MSGF:ScoreRatio",
+    "MSGF:Energy",
+    "MSGF:lnEValue",
+    "MSGF:lnExplainedIonCurrentRatio",
+    "MSGF:lnNTermIonCurrentRatio",
+    "MSGF:lnCTermIonCurrentRatio",
+    "MSGF:lnMS2IonCurrent",
+    "MSGF:MeanErrorTop7",
+    "MSGF:sqMeanErrorTop7",
+    "MSGF:StdevErrorTop7",
+    "XTANDEM:hyperscore",
+    "XTANDEM:deltascore",
+    "MS:1001330",  # expect_score
+    "hyperscore",  # MSfragger
+    "nextscore",  # MSfragger
+    "COMET:deltaCn",  # recalculated deltaCn = (current_XCorr - 2nd_best_XCorr) / max(current_XCorr, 1)
+    "COMET:deltaLCn",  # deltaLCn = (current_XCorr - worst_XCorr) / max(current_XCorr, 1)
+    "COMET:lnExpect",  # log(E-value)
+    "MS:1002252",  # unchanged XCorr
+    "MS:1002255",  # unchanged Sp = number of candidate peptides
+    "COMET:lnNumSP",  # log(number of candidate peptides)
+    "COMET:lnRankSP",  # log(rank based on Sp score)
+    "COMET:IonFrac",  # matched_ions / total_ions
+    "MS:1001171",  # unchanged mScore
+    "MASCOT:delta_score",  # delta score based on mScore
+    "CONCAT:lnEvalue",
+    "CONCAT:deltaLnEvalue",
+    "SAGE:ln(-poisson)"
+    "SAGE:ln(delta_best)",
+    "SAGE:ln(delta_next)",
+    "SAGE:ln(matched_intensity_pct)",
+    "SAGE:longest_b",
+    "SAGE:longest_y",
+    "SAGE:longest_y_pct",
+    "SAGE:matched_peaks",
+    "SAGE:scored_candidates"
+]
 
 class IdXMLReader(ReaderBase):
-    def __iter__(self):
-        """Iterate over file and return PSMs one-by-one."""
-        with idxml.read(str(self.filename)) as reader:
-            for entry in reader:
-                for peptide_hit in entry["PeptideHit"]:
-                    yield self._parse_psm(entry, peptide_hit)
+    def __init__(self, filename: str | Path, *args, **kwargs) -> None:
```

Comment:
todo: Replace union operator with the imported `Union` class from `typing`. For backwards compatibility, we are stuck with these instead of the nice operator syntax until Python 3.9 is end-of-life in 2026 :(